mlos_core/mlos_core/optimizers/bayesian_optimizers/smac_optimizer.py

```diff
@@ -257,20 +305,68 @@ def _register(self, *, configs: pd.DataFrame,
             Not Yet Implemented.
 
         metadata: pd.DataFrame
-            Not Yet Implemented.
+            The metadata for the config to register.
         """
-        from smac.runhistory import StatusType, TrialInfo, TrialValue  # pylint: disable=import-outside-toplevel
+        from smac.runhistory import StatusType, TrialValue  # pylint: disable=import-outside-toplevel
 
         if context is not None:
             warn(f"Not Implemented: Ignoring context {list(context.columns)}", UserWarning)
 
-        # Register each trial (one-by-one)
-        for (config, (_i, score)) in zip(self._to_configspace_configs(configs=configs), scores.iterrows()):
-            # Retrieve previously generated TrialInfo (returned by .ask()) or create new TrialInfo instance
-            info: TrialInfo = self.trial_info_map.get(
-                config, TrialInfo(config=config, seed=self.base_optimizer.scenario.seed))
-            value = TrialValue(cost=list(score.astype(float)), time=0.0, status=StatusType.SUCCESS)
-            self.base_optimizer.tell(info, value, save=False)
+        with self.lock:
+            # Register each trial (one-by-one)
+            metadatas: Union[List[pd.Series], List[None]] = to_metadata(metadata) or [
+                None for _ in scores   # type: ignore[misc]
+            ]
+            for config, score, ctx in zip(
+                self._to_configspace_configs(configs=configs),
+                scores.values.tolist(),
+                metadatas,
+            ):
+                value: TrialValue = TrialValue(
+                    cost=score, time=0.0, status=StatusType.SUCCESS
+                )
+
+                matching: pd.Series[bool]
+                if ctx is None:
+                    matching = self.trial_info_df["Configuration"] == config
+                else:
+                    matching = (
+                        self.trial_info_df["Configuration"] == config
+                    ) & pd.Series(
+                        [df_ctx.equals(ctx) for df_ctx in self.trial_info_df["Metadata"]]
+                    )
+
+                # make a new entry
```

Comment:
Comments could use some improvement.

Make a new entry of what?  What's the first part doing that this one now needs explanation?